internal/cache/cache.go

```diff
@@ -0,0 +1,113 @@
+package cache
+
+import (
+	"sync"
+	"time"
+)
+
+// Cache defines the interface for cache operations
+type Cache[K comparable, V any] interface {
```

Comment:
What about an off the shelf library?
https://github.com/eko/gocache
https://github.com/allegro/bigcache
https://github.com/dgraph-io/ristretto
https://github.com/patrickmn/go-cache

Reply:
I checked the libs you provided and they look quite advanced. I'm not sure about our future requirements but our current needs are quite simple. It might makes sense to go homemade for our usecase.

Do you have any experience with those libs? What do you recommend?

Reply:
I almost used ristretto for a project that needed high performance, but you're probably right that it's excessive for our needs.

Two trains of thought right now:
1. `go-cache` seems like an excellent fit. It's extremely lightweight (a single file with ~1000 LOC), the interface is almost identical to the one you have, the project has the most stars.

2. If we do roll our own, I bet we could come up with a clever optimization to time based expiration utilizing sequence numbers and multiple maps.

Reply:
We can go with a with `go-cache` and overload it with a custom logic to implement the `customPolicy`. Something like this:

```
type CustomCache[K comparable, V any] struct {
	cache        *cache.Cache
	customPolicy func(V) bool
	mutex        sync.RWMutex
}

func (c *CustomCache[K, V]) Get(key K) (V, bool) {
	c.mutex.RLock()
	defer c.mutex.RUnlock()

        ...

	// Check custom policy
	if c.customPolicy != nil && c.customPolicy(typedValue) {
		c.cache.Delete(interface{}(key))
		return zero, false
	}

	return typedValue, true
}
```